### ISSUE TITLE
App crashes when subscribing to a topic

### ISSUE THREAD

**Issue by lukin0110** [aws/aws-iot-device-sdk-js#109]
I have a device with a bad internet connection. We the connection drops it tries to reconnect. This is a piece of code that we're already using for more then a year. We've updated the SDK to 1.0.14 two weeks ago.
 
Since today we receive this error once in a while and it causes a crash of our node app:
```
Process uncaughtException: TypeError: Cannot read property 'length' of undefined
     at /app/node_modules/aws-iot-device-sdk/thing/index.js:174:52
     at that.outgoing.(anonymous function) (/app/node_modules/mqtt/lib/client.js:138:15)
     at /app/node_modules/mqtt/lib/client.js:845:9
     at Store.del (/app/node_modules/mqtt/lib/store.js:76:5)
     at MqttClient._handleAck (/app/node_modules/mqtt/lib/client.js:831:26)
     at MqttClient._handlePacket (/app/node_modules/mqtt/lib/client.js:284:12)
     at process (/app/node_modules/mqtt/lib/client.js:233:12)
     at Writable.writable._write (/triflux/node_modules/mqtt/lib/client.js:243:5)
     at doWrite (/app/node_modules/readable-stream/lib/_stream_writable.js:237:10)
     at writeOrBuffer (/app/node_modules/readable-stream/lib/_stream_writable.js:227:5)
```

**Comment by fengsongAWS** [comment]
Hi @lukin0110 ,
Thanks for your interest in AWS IoT Device SDK.
I am trying to duplicate your issues. It would be more helpful if you provide more details of your use case and settings.
The error itself indicates the variable granted is passed incorrectly as shown here https://github.com/aws/aws-iot-device-sdk-js/blob/master/thing/index.js#L174 From what I see, this only happens to subscribe. Could you try to identify the trace route of this error? Like does this happen during the reconnect or during shadow register or the first time when you try to connect? Are you using persistent subscribe or non-persistent subscribe? How frequently does it happen to you?

**Comment by SubaruWRX55** [comment]
I'm getting same error. Happens when the connection is unstable.

Connection d404c7f53a42c22fb8421db764b5d9-3 has connected to AWS-iot
Feb 27 16:09:56 hub1 rentlyhub.js[370]: Connection d404c7f53a42c22fb8421db764b5d9-3 offline
Feb 27 16:09:56 hub1 rentlyhub.js[370]: Connection d404c7f53a42c22fb8421db764b5d9-3 has been closed
Feb 27 16:09:57 hub1 rentlyhub.js[370]: Connection d404c7f53a42c22fb8421db764b5d9-3 attempts to reconnect to AWS-iot
Feb 27 16:09:58 hub1 rentlyhub.js[370]: Connection d404c7f53a42c22fb8421db764b5d9-3 has connected to AWS-iot
Feb 27 16:09:58 hub1 rentlyhub.js[370]: Caught unexpected exception TypeError: Cannot read property 'length' of undefined
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at /root/hub/node_modules/aws-iot-device-sdk/thing/index.js:174:52
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at that.outgoing.(anonymous function) (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/client.js:138:15)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at /root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/client.js:845:9
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at Store.del (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/store.js:76:5)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at MqttClient._handleAck (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/client.js:831:26)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at MqttClient._handlePacket (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/client.js:284:12)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at process (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/client.js:233:12)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at Writable.writable._write (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/lib/client.js:243:5)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at doWrite (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/node_modules/readable-stream/lib/_stream_writable.js:237:10)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at writeOrBuffer (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/node_modules/readable-stream/lib/_stream_writable.js:227:5)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at Writable.write (/root/hub/node_modules/aws-iot-device-sdk/node_modules/mqtt/node_modules/readable-stream/lib/_stream_writable.js:194:11)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at TLSSocket.ondata (_stream_readable.js:528:20)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at emitOne (events.js:77:13)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at TLSSocket.emit (events.js:169:7)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at readableAddChunk (_stream_readable.js:146:16)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at TLSSocket.Readable.push (_stream_readable.js:110:10)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: at TLSWrap.onread (net.js:529:20)
Feb 27 16:09:58 hub1 rentlyhub.js[370]: This error cannot be handled now, reboot...

**Comment by fengsongAWS** [comment]
Hi @SubaruWRX55 ,
Does this happen all the time when trying to reconnect? Could you elaborate more about your use cases?
Thanks

**Comment by SubaruWRX55** [comment]
When I created and maintained 4 iot connections per hub, this error starts to occur very frequently. Average 30 minute each time it happens.

**Comment by fengsongAWS** [comment]
Fixed in v2.0.0
